add cpu feature detection for Arm64

lubinsz opened the pull request:
Signed-off-by: Bin Lu <bin.lu@arm.com>

lubinsz commented:
@klauspost 
Hi @klauspost ,

Why Travis report this issue?
Log:
-----------------------------------------------------------
The command "go test -race ./..." exited with 0.
0.13s$ diff <(gofmt -d .) <("")
: command not found
------------------------------------------------------------
It seems that no command was found in Travis. Is there any problem in your travis test script?

klauspost commented:
The problem is that your code doesn't appear to be go fmt'ed. Is that correct?

lubinsz commented:
Hi @klauspost 
Please take a look at this pr. 

With this patch, I enabled feature detection for Arm64 platform. 
Thanks.

klauspost commented on detect_arm64.go:
```diff
@@ -0,0 +1,246 @@
+ */
+func armCpuid(op uint32) (eax, ebx, ecx, edx uint32) {
+	eax = uint32(cpuInfo_arm64.HwCap)
+	ebx = 0
```
These are always 0, no need to set them

klauspost commented on detect_arm64.go:
```diff
@@ -0,0 +1,246 @@
+// Copyright (c) 2018 Robin Lu (lubinsz@gmail.com), released under MIT License. See LICENSE file.
+// +build arm64,linux
```
AFAICT `cgo` should also be added as a build constraint.

klauspost commented on detect_arm64.go:
```diff
@@ -0,0 +1,246 @@
+	cpuInfo_arm64 *cpuinfo_arm
+)
+
+func CPUInit_arm() {
```
This should not be exported.

klauspost commented on cpuid.go:
```diff
@@ -178,6 +235,17 @@ func init() {
+// isArm64Arch returns true if the architecture is ARM64.
+//
+// The arch value should be normalized before being passed to this function.
+func isArm64Arch() bool {
```
It should be possible to make this static, ie. set by a build constraint.

lubinsz commented:
Hi @klauspost 
a) I would like to expose the boolean value returned by isArm64Arch. Maybe by adding ARM64 as a "Vendor".
-------------
The vendor for arm64 should be Cavium, Qualcomm, Marvell..
I will update the vendor for Arm64 in another patch.

Also can I add "cpuFamily" in the struct of "CPUInfo" ?

b) Is is possible to populate more of "CPU" variable?
----
In Arm64 platform, all "CPU" variables can be get by 'midr register' & 'hwcap'.
Please see link(http://infocenter.arm.com/help/topic/com.arm.doc.100235_0003_00_en/arm_cortex_m33_dgug_100235_0003_00_en.pdf) as reference.
Such as: 'CPU implementer' in 'midr register' means cpu vendor.
I have updated detect_arm64.go, please check it.

c) Please add tests for valid/invalid different inputs to "decodeCpuInfo". I really don't want this to blow up.
-----
I have removed it

d) Please add some documentation/examples on how to use it.
-----
I will update it in another patch :)

lubinsz commented:
Hi @klauspost ,
Can you take a look at it？
Thanks

lubinsz commented:
Hi @klauspost ,
Can you take a look at it？
Thanks

fwessels commented:
Superseded by PR #44
